src/gt4py/frontend/defir_to_gtir.py

```diff
@@ -81,6 +85,171 @@ def _make_literal(v: numbers.Number) -> gtir.Literal:
+        name = node.name 
+        if params[name].data_dims:
+            field_list = []
+            # vector
+            if len(params[name].data_dims) == 1:
+                dims = params[name].data_dims[0]
+                for index in range(dims):
+                    tmp = copy.deepcopy(node)
+                    data_type = DataType.INT32
+                    data_index = [ScalarLiteral(value=index, data_type=data_type)]
+                    tmp.data_index = data_index
+                    field_list.append(tmp)
+            # matrix
+            elif len(params[name].data_dims) == 2:
+                rows, cols = params[name].data_dims
+                for row in range(rows):
+                    row_list = []
+                    for col in range(cols):
+                        tmp = copy.deepcopy(node)
+                        data_type = DataType.INT32
+                        data_index = [ScalarLiteral(value=row, data_type=data_type), ScalarLiteral(value=col, data_type=data_type)]
+                        tmp.data_index = data_index
+                        row_list.append(tmp)
+                    field_list.append(row_list)
+            else:
+                raise Exception("Higher dimensional fields only supported for vectors and matrices")
+
+        return field_list
+
+    def add_node(self, lhs_node, rhs_node, binary_op: BinaryOperator, loc: Location):
+        return BinOpExpr(op=binary_op, lhs=lhs_node, rhs=rhs_node, loc=loc)
+
+    def visit_UnaryOpExpr(self, node: UnaryOpExpr, params):
+        # assert node.op == UnaryOperator.TRANSPOSED, f'Unsupported unary operator {node.op}'
+        if node.op == UnaryOperator.TRANSPOSED:
+            node = self.visit(node.arg, params=params)
+            # transpose list
+            node = [list(x) for x in zip(*node)]
+            return node
+            
+        else:
+            return node
+
+    def visit_BinOpExpr(self, node: BinOpExpr, params):
+
+        loc = node.loc
+        lhs_list = self.visit(node.lhs, params=params)
+        rhs_list = self.visit(node.rhs, params=params)
+        bin_op_list = []
+
+        if node.op == BinaryOperator.MATMULT:
+            for j in range(len(lhs_list)):
+                acc = self.add_node(lhs_list[j][0], rhs_list[0], BinaryOperator.MUL, loc)
+                for i in range(1, len(lhs_list[0])):
+                    mul = self.add_node(lhs_list[j][i], rhs_list[i], BinaryOperator.MUL, loc)
+                    acc = self.add_node(acc, mul, BinaryOperator.ADD, loc)
+
+                bin_op_list.append(acc)
+        
+        else:
+            # both vectors
+            if type(lhs_list) == list and type(rhs_list) == list:
+                if len(lhs_list) == len(rhs_list):
+                    for lhs, rhs in zip(lhs_list, rhs_list):
+                        bin_op_list.append(self.add_node(lhs, rhs, node.op, node.loc))
+            # scalar and vector
+            elif type(rhs_list) == list:
```

Comment:
```suggestion
            elif isinstance(lhs_list, Expr) and isinstance(rhs_list, list):
```
The previous branch already ensures the lhs to be an expression, but I would suggest this more literate condition (same applies to the branches below).